### ISSUE TITLE
bridge: Not able to use in a custom network namspace

### ISSUE THREAD

**Issue by alxbse** [containernetworking/plugins#419]
I would like to be able to use the bridge plugin and have it create/use a bridge in a custom network namespace. Configuration wise I would like to do something like this:

```json
{
	"name": "mynet",
	"type": "bridge",
	"bridge": "mynet0",
        "namespace": "mynamespace",
	"ipam": {}
}
```

This would create the bridge in the network namespace `mynamespace` rather than the host network namespace.

Is this something that could be added to the bridge plugin or is it an intentional limitation?

**Comment by squeed** [comment]
It's not a bad idea, though can you describe the use-case?

FWIW, you can do this now just by executing the CNI binary in the desired "host" netns - it doesn't explicitly discover it. Now, k8s won't let you do this, of course...

**Comment by alxbse** [comment]
My use case is a kubernetes project where all container hosts have two network interfaces. One for host services and one intended for pod traffic. All routes the pods are intended to use are announced via BGP but since the host services and pods ends up using the same routing table host services tries to reach resources external to the cluster through routes intended only for the pods.

My use case can be solved by creating custom routing tables and rules but a more elegant solution I have experimented with is to have a dedicated namespace for the CNI bridge. Moving the second interface and the BGP solution into the dedicated namespace gives me a nice clean separation of traffic without any need for manually create routing rules.

Hopefully that gives a bit more info on my use case. I have a local branch of the bridge plugin with namespace support to verify the concept and would be happy to create a pull request from it. I just want to verify that it's not out of scope for this project or if there is maybe another simple way of solving it.

**Comment by mars1024** [comment]
IMO, this should be the job of CRI or something like CNIManager to decide the "host" namespace.

**Comment by alxbse** [comment]
@mars1024 By CNIManager you mean a solution like Calico or Weave?

**Comment by mars1024** [comment]
I mean something like `cniNetworkPlugin` in `k8s.io/kubernetes/pkg/kubelet/dockershim/cni`.

**Comment by Thoro** [comment]
I do have another use case, which cost me quite some time to figure out.

This only affects routed network solutions on kubernetes (kube-router), with firewalling on bridge level (custom), and NAT  ipvs (kube-router).

If a bridge is firewalled via iptables using net.bridge.bridge-nf-call-iptables = 1 then the POSTROUTING table is called only ONCE for the packets. This leads to an impossible state if the local pod is calling a local service in ipvs because the outgoing packets to other pods cannot be masqueraded anymore:

10.253.1.0/24 = pod network 1
10.253.2.0/24 = pod network 2
10.253.1.1 = gateway host 1
10.10.0.10 = ipvs service

iptables-rules:

iptables -t nat -A POSTROUTING -m ipvs -vdir ORIGINAL -j MASQUERADE

sample packet flow:

10.253.1.10 -> 10.10.0.10 (inside of bridge) (hits in order: BRIDGE FORWARD, BRIDGE POSTROUTING, IPVS)
10.253.1.10 -> 10.253.2.5 (outgoing packets) (this packet should get SNAT, but can't since POSTROUTING was already hit)
10.253.2.5 -> 10.253.1.10  (returning packet) (now this packet cannot be routed correctly, because it's returned directly to the pod)


So the problem is, you can only do SNAT on POSTROUTING, but IPVS takes the packet only AFTER POSTROUTING of the bridge, then the packet, since it was already in POSTROUTING will never get there again.

The only solution to this is to move the bridge into a separate network namespace, and actually use a veth pair to connect to this bridge from the host. (since brouting will directly forward the incoming packets from eth to bridge, without specifing the bridge interfaces (i.e. -m physdev --phys-dev-out DOES NOT work))

Might be a very specific use case, and I'm also not sure if it will break some expectation regarding "host" namespace.
